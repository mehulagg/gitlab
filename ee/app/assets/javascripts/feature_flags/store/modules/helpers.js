import _ from 'underscore';
import {
  ROLLOUT_STRATEGY_ALL_USERS,
  ROLLOUT_STRATEGY_PERCENT_ROLLOUT,
  INTERNAL_ID_PREFIX,
  DEFAULT_PERCENT_ROLLOUT,
  PERCENT_ROLLOUT_GROUP_ID,
} from '../../constants';

/**
 * Converts raw scope objects fetched from the API into an array of scope
 * objects that is easier/nicer to bind to in Vue.
 * @param {Array} scopesFromRails An array of scope objects fetched from the API
 */
export const mapToScopesViewModel = scopesFromRails =>
  (scopesFromRails || []).map(s => {
    const [strategy] = s.strategies || [];

    const rolloutStrategy = strategy ? strategy.name : ROLLOUT_STRATEGY_ALL_USERS;

    let rolloutPercentage = DEFAULT_PERCENT_ROLLOUT;
    if (strategy && strategy.parameters && strategy.parameters.percentage) {
      rolloutPercentage = strategy.parameters.percentage;
    }

    return {
      id: s.id,
      environmentScope: s.environment_scope,
      active: Boolean(s.active),
      canUpdate: Boolean(s.can_update),
      protected: Boolean(s.protected),
      rolloutStrategy,
      rolloutPercentage,

      // eslint-disable-next-line no-underscore-dangle
      shouldBeDestroyed: Boolean(s._destroy),
    };
  });

/**
 * Converts the parameters emitted by the Vue component into
 * the shape that the Rails API expects.
 * @param {Array} scopesFromVue An array of scope objects from the Vue component
 */
export const mapFromScopesViewModel = params => {
  const scopes = (params.scopes || []).map(s => {
    const parameters = {};
    if (s.rolloutStrategy === ROLLOUT_STRATEGY_PERCENT_ROLLOUT) {
      parameters.groupId = PERCENT_ROLLOUT_GROUP_ID;
      parameters.percentage = s.rolloutPercentage;
    }

    // Strip out any internal IDs
    const id = _.isString(s.id) && s.id.startsWith(INTERNAL_ID_PREFIX) ? undefined : s.id;

    return {
      id,
      environment_scope: s.environmentScope,
      active: s.active,
      can_update: s.canUpdate,
      protected: s.protected,
      _destroy: s.shouldBeDestroyed,
      strategies: [
        {
          name: s.rolloutStrategy,
          parameters,
        },
      ],
    };
  });

  return {
    operations_feature_flag: {
      name: params.name,
      description: params.description,
      scopes_attributes: scopes,
    },
  };
};

/**
 * Creates a new feature flag environment scope object for use
 * in a Vue component.  An optional parameter can be passed to
 * override the property values that are created by default.
 *
 * @param {Object} overrides An optional object whose
 * property values will be used to override the default values.
 *
 */
export const createNewEnvironmentScope = (overrides = {}) => {
  const defaultScope = {
    environmentScope: '',
    active: false,
    id: _.uniqueId(INTERNAL_ID_PREFIX),
    rolloutStrategy: ROLLOUT_STRATEGY_ALL_USERS,
    rolloutPercentage: DEFAULT_PERCENT_ROLLOUT,
  };

  const newScope = {
    ...defaultScope,
    ...overrides,
  };

  if (gon && gon.features && gon.features.featureFlagPermissions) {
    newScope.canUpdate = true;
    newScope.protected = false;
  }

  return newScope;
};
