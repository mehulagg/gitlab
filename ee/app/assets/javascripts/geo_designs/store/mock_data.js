import { __ } from '~/locale';

function generateRandomDateString() {
  return new Date(
    new Date(2019, 0, 1).getTime() +
      Math.random() * (new Date().getTime() - new Date(2019, 0, 1).getTime()),
  ).toString();
}

const mockData = [
  {
    id: 1,
    name: __("Zack's Design Repo"),
    url: 'http://localhost:3002',
    sync_status: 'synced',
    last_synced_at: generateRandomDateString(),
    last_verified_at: generateRandomDateString(),
    last_checked_at: generateRandomDateString(),
  },
  {
    id: 2,
    name: __("Valery's Design Repo"),
    url: 'http://localhost:3002',
    sync_status: 'pending',
    last_synced_at: generateRandomDateString(),
    last_verified_at: generateRandomDateString(),
    last_checked_at: generateRandomDateString(),
  },
  {
    id: 3,
    name: __("Mike's Design Repo"),
    url: 'http://localhost:3002',
    sync_status: 'failed',
    last_synced_at: generateRandomDateString(),
    last_verified_at: generateRandomDateString(),
    last_checked_at: generateRandomDateString(),
  },
  {
    id: 4,
    name: __("Rachel's Design Repo"),
    url: 'http://localhost:3002',
    sync_status: null,
    last_synced_at: null,
    last_verified_at: null,
    last_checked_at: null,
  },
];

export default mockData;
