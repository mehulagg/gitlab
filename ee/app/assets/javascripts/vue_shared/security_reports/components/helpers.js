import { EMPTY_BODY_MESSAGE } from './constants';

export const bodyWithFallBack = (body) => {
  // Return string if defined or blank string, otherwise fallback
  return body === '' ? '' : body || EMPTY_BODY_MESSAGE;
};
