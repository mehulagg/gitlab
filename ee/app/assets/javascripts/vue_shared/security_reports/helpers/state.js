export default {
  sast: {
    isLoading: false,
    hasError: false,
    newIssues: [],
    resolvedIssues: [],
    allIssues: [],
  },
  sastContainer: {
    approved: [],
    unapproved: [],
    vulnerabilities: [],
  },
  dast: [],
  codeclimate: {
    newIssues: [],
    resolvedIssues: [],
  },
};
