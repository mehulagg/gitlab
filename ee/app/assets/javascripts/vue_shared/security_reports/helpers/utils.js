import { stripHtml } from '~/lib/utils/text_utility';

/**
 * Parses SAST and Codeclimate Issues into a common and reusable format
 * to reuse the same vue component.
 * [
 *   {
 *     name: String,
 *     priority: String,
 *     fingerprint: String,
 *     path: String,
 *     line: Number,
 *     urlPath: String
 *   }
 * ]
 * @param {array} issues
 * @return {array}
 */
export const parseIssues = (issues = [], path = '') => issues.map((issue) => {
  const parsedIssue = {
    name: issue.description || issue.message,
    ...issue,
  };

  // code quality
  if (issue.location) {
    let parseCodeQualityUrl;

    if (issue.location.path) {
      parseCodeQualityUrl = `${path}/${issue.location.path}`;
      parsedIssue.path = issue.location.path;
    }

    if (issue.location.lines && issue.location.lines.begin) {
      parsedIssue.line = issue.location.lines.begin;
      parseCodeQualityUrl += `#L${issue.location.lines.begin}`;
    }

    parsedIssue.urlPath = parseCodeQualityUrl;

  // security
  } else if (issue.file) {
    let parsedSecurityUrl = `${path}/${issue.file}`;
    parsedIssue.path = issue.file;

    if (issue.line) {
      parsedSecurityUrl += `#L${issue.line}`;
    }
    parsedIssue.urlPath = parsedSecurityUrl;
  }

  return parsedIssue;
});

/**
 * Compares two arrays by the given key and returns the difference
 *
 * @param {Array} firstArray
 * @param {Array} secondArray
 * @param {String} key
 * @returns {Array}
 */
export const filterByKey = (firstArray = [], secondArray = [], key = '') => firstArray
  .filter(item => !secondArray.find(el => el[key] === item[key]));

/**
 * Parses DAST results into a common format to allow to use the same Vue component
 * And adds an external link
 *
 * @param {Array} data
 * @returns {Array}
 */
export const parseSastContainer = (data = []) => data.map(el => ({
  name: el.vulnerability,
  priority: el.severity,
  path: el.namespace,
  // external link to provide better description
  nameLink: `https://cve.mitre.org/cgi-bin/cvename.cgi?name=${el.vulnerability}`,
  ...el,
}));

/**
 * Utils functions to set the reports
 */

/**
 * Compares sast results and returns the formatted report
 *
 * Security report has 3 types of issues, newIssues, resolvedIssues and allIssues.
 *
 * When we have both base and head:
 * - newIssues = head - base
 * - resolvedIssues = base - head
 * - allIssues = head - newIssues - resolvedIssues
 *
 * When we only have head
 * - newIssues = head
 * - resolvedIssues = 0
 * - allIssues = 0
 * @param {*} data
 * @returns {Object}
 */
export const setSastReport = (data = {}) => {
  const securityReport = {};

  if (data.base) {
    const filterKey = 'cve';
    const parsedHead = parseIssues(data.head, data.headBlobPath);
    const parsedBase = parseIssues(data.base, data.baseBlobPath);

    securityReport.newIssues = filterByKey(
      parsedHead,
      parsedBase,
      filterKey,
    );
    securityReport.resolvedIssues = filterByKey(
      parsedBase,
      parsedHead,
      filterKey,
    );

    // Remove the new Issues and the added issues
    securityReport.allIssues = filterByKey(
      parsedHead,
      securityReport.newIssues.concat(securityReport.resolvedIssues),
      filterKey,
    );
  } else {
    securityReport.newIssues = parseIssues(data.head, data.headBlobPath);
  }

  return securityReport;
};

export const setSastContainerReport = (data = {}) => {
  const unapproved = data.unapproved || [];

  const parsedVulnerabilities = parseSastContainer(data.vulnerabilities);

  // Approved can be calculated by subtracting unapproved from vulnerabilities.
  return {
    vulnerabilities: parsedVulnerabilities || [],
    approved: parsedVulnerabilities
    .filter(item => !unapproved.find(el => el === item.vulnerability)) || [],
    unapproved: parsedVulnerabilities
    .filter(item => unapproved.find(el => el === item.vulnerability)) || [],
  };
};

/**
 * Dast Report sends some keys in HTML, we need to strip the `<p>` tags.
 * This should be moved to the backend.
 *
 * @param {Array} data
 * @returns {Array}
 */
export const setDastReport = data => data.site.alerts.map(alert => ({
  name: alert.name,
  parsedDescription: stripHtml(alert.desc, ' '),
  priority: alert.riskdesc,
  ...alert,
}));
