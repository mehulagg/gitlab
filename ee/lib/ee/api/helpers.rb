module EE
  module API
    module Helpers
      extend ::Gitlab::Utils::Override

      override :current_user
      def current_user
        strong_memoize(:current_user) do
          user = super

          if user
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :user, user.id)
          end

          user
        end
      end

      def check_project_feature_available!(feature)
        not_found! unless user_project.feature_available?(feature)
      end
    end
  end
end
