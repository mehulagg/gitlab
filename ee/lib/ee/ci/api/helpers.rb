module EE
  module Ci
    module API
      module Helpers
        def authenticate_build!
          id = params[:id]

          if id
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :build, id)
          end

          super
        end

        def current_runner
          token = params[:token]

          if token
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :runner, token)
          end

          super
        end
      end
    end
  end
end
