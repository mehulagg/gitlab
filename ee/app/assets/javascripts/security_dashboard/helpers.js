import isPlainObject from 'lodash/isPlainObject';
import { ALL, BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { s__, __ } from '~/locale';

const parseOptions = obj =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = projects =>
  projects.map(p => ({ id: p.id.split('/').pop(), name: p.name }));

export const stateFilter = {
  name: s__('SecurityReports|Status'),
  id: 'state',
  allOption: { id: ALL, name: s__('VulnerabilityStatusTypes|All') },
  options: parseOptions(VULNERABILITY_STATES),
};

export const severityFilter = {
  name: s__('SecurityReports|Severity'),
  id: 'severity',
  allOption: BASE_FILTERS.severity,
  options: parseOptions(SEVERITY_LEVELS),
};

const getScannerFilterOptions = () => {
  const options = parseOptions(REPORT_TYPES);
  return options.map(option => ({
    ...option,
    queryId: `GitLab:${option.id}`,
    externalId: 'NO_SCANNERS_FOUND',
  }));
};

export const scannerFilter = {
  name: s__('Reports|Scanner'),
  id: 'reportType',
  allOption: BASE_FILTERS.report_type,
  options: getScannerFilterOptions(),
};

export const getProjectFilter = projects => {
  if (!projects?.length) return undefined;

  return {
    name: s__('SecurityReports|Project'),
    id: 'projectId',
    allOption: BASE_FILTERS.project_id,
    options: mapProjects(projects),
  };
};

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter(entry => entry !== null);
};

/**
 * We have disabled loading hasNextPage from GraphQL as it causes timeouts in database,
 * instead we have to calculate that value based on the existence of endCursor. When endCursor
 * is empty or has null value, that means that there is no next page to be loaded from GraphQL API.
 *
 * @param {Object} pageInfo
 * @returns {Object}
 */
export const preparePageInfo = pageInfo => {
  return { ...pageInfo, hasNextPage: Boolean(pageInfo?.endCursor) };
};

export const createProjectLoadingError = () => __('An error occurred while retrieving projects.');

export default () => ({});
