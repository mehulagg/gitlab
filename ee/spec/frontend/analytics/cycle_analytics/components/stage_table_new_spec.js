import { GlEmptyState, GlLoadingIcon } from '@gitlab/ui';
import { shallowMount, mount } from '@vue/test-utils';
import StageTableNew from 'ee/analytics/cycle_analytics/components/stage_table_new.vue';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import { issueEvents, issueStage, testStage } from '../mock_data';

let wrapper = null;

const noDataSvgPath = 'path/to/no/data';
const emptyStateMessage = 'Too much data';
const notEnoughDataError = "We don't have enough data to show this stage.";

const findStageEvents = () => wrapper.findAllByTestId('vsa-stage-event');
const findStageEventTitle = (ev) => extendedWrapper(ev).findByTestId('vsa-stage-event-title');

function createComponent(props = {}, shallow = false) {
  const func = shallow ? shallowMount : mount;
  return extendedWrapper(
    func(StageTableNew, {
      propsData: {
        isLoading: false,
        stageEvents: issueEvents,
        noDataSvgPath,
        currentStage: issueStage,
        ...props,
      },
      stubs: {
        'gl-loading-icon': true,
        'gl-empty-state': true,
      },
    }),
  );
}

describe('StageTable', () => {
  afterEach(() => {
    wrapper.destroy();
  });

  describe('is loaded with data', () => {
    beforeEach(() => {
      wrapper = createComponent();
    });

    it('will render the correct events', () => {
      const evs = findStageEvents();
      expect(evs).toHaveLength(issueEvents.length);

      const titles = evs.wrappers.map((ev) => findStageEventTitle(ev).text());
      issueEvents.forEach((ev, index) => {
        expect(titles[index]).toBe(ev.title);
      });
    });

    it('will not display the default data message', () => {
      expect(wrapper.html()).not.toContain(notEnoughDataError);
    });
  });

  it('will render an event with an iid', () => {
    const [firstEvent] = issueEvents;
    wrapper = createComponent({ stageEvents: [firstEvent] });
    expect(wrapper.findByTestId('vsa-stage-event-build-status').exists()).toBe(false);
    expect(wrapper.findByTestId('vsa-stage-event').html()).toMatchSnapshot();
  });

  it('will render an event with an id', () => {
    const [firstEvent] = issueEvents;
    wrapper = createComponent({
      stageEvents: [{ ...firstEvent, id: 222 }],
      currentStage: { ...issueStage, custom: false },
    });
    expect(wrapper.findByTestId('vsa-stage-event-build-status').exists()).toBe(false);
    expect(wrapper.findByTestId('vsa-stage-event').html()).toMatchSnapshot();
  });

  it('will render build status for the default `test` stage', () => {
    const [firstEvent] = issueEvents;
    wrapper = createComponent({
      stageEvents: [{ ...firstEvent, id: 222 }],
      currentStage: { ...testStage, custom: false },
    });
    expect(wrapper.findByTestId('vsa-stage-event-build-status').exists()).toBe(true);
    expect(wrapper.findByTestId('vsa-stage-event').html()).toMatchSnapshot();
  });

  it('isLoading = true', () => {
    wrapper = createComponent({ isLoading: true }, true);
    expect(wrapper.find(GlLoadingIcon).exists()).toEqual(true);
  });

  describe('with no stageEvents', () => {
    beforeEach(() => {
      wrapper = createComponent({ stageEvents: [] });
    });

    it('will render the empty state', () => {
      expect(wrapper.find(GlEmptyState).exists()).toBe(true);
    });

    it('will display the default no data message', () => {
      expect(wrapper.html()).toContain(notEnoughDataError);
    });
  });

  describe('emptyStateMessage set', () => {
    beforeEach(() => {
      wrapper = createComponent({ stageEvents: [], emptyStateMessage });
    });

    it('will display the custom message', () => {
      expect(wrapper.html()).not.toContain(notEnoughDataError);
      expect(wrapper.html()).toContain(emptyStateMessage);
    });
  });
});
