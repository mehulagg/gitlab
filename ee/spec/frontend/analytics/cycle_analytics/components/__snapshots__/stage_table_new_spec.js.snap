// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StageTable will render an event with an id 1`] = `
"<div data-testid=\\"vsa-stage-event\\">
  <div data-testid=\\"vsa-stage-content\\">
    <p class=\\"gl-m-0\\">
      <!----> <a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal pipeline-id\\">#222</a> <svg data-testid=\\"fork-icon\\" aria-hidden=\\"true\\" class=\\"gl-icon s16\\">
        <use href=\\"#fork\\"></use>
      </svg>
      <!----> <span class=\\"icon-branch gl-text-gray-400\\"><svg data-testid=\\"commit-icon\\" aria-hidden=\\"true\\" class=\\"gl-icon s14\\"><use href=\\"#commit\\"></use></svg></span> <a href=\\"#\\" class=\\"gl-link commit-sha\\"></a></p>
    <p class=\\"gl-m-0\\"><span><a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal build-date\\"></a>
              by
              <a href=\\"http://localhost/user1\\" class=\\"gl-link gl-text-black-normal issue-author-link\\">John Doe2</a></span></p>
  </div>
</div>"
`;

exports[`StageTable will render an event with an iid 1`] = `
"<div data-testid=\\"vsa-stage-event\\">
  <div data-testid=\\"vsa-stage-content\\">
    <h5 data-testid=\\"vsa-stage-event-title\\" class=\\"gl-font-weight-bold gl-my-1\\"><a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal\\">My title 10</a></h5>
    <p class=\\"gl-m-0\\"><a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal\\">#3</a> <span class=\\"gl-font-lg\\">·</span> <span>
              Opened
              <a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal\\">4 days ago</a></span> <span>
              by
              <a href=\\"http://localhost/user1\\" class=\\"gl-link gl-text-black-normal\\">John Doe2</a></span></p>
  </div>
</div>"
`;

exports[`StageTable will render build status for the default \`test\` stage 1`] = `
"<div data-testid=\\"vsa-stage-event\\">
  <div data-testid=\\"vsa-stage-content\\">
    <p class=\\"gl-m-0\\"><span data-testid=\\"vsa-stage-event-build-status\\" class=\\"icon-build-status gl-vertical-align-middle gl-text-green-500\\"><svg data-testid=\\"status_success-icon\\" aria-hidden=\\"true\\" class=\\"gl-icon s14\\"><use href=\\"#status_success\\"></use></svg></span> <a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal item-build-name\\">

      </a>
      ·
      <a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal pipeline-id\\">#222</a> <svg data-testid=\\"fork-icon\\" aria-hidden=\\"true\\" class=\\"gl-icon s16\\">
        <use href=\\"#fork\\"></use>
      </svg>
      <!----> <span class=\\"icon-branch gl-text-gray-400\\"><svg data-testid=\\"commit-icon\\" aria-hidden=\\"true\\" class=\\"gl-icon s14\\"><use href=\\"#commit\\"></use></svg></span> <a href=\\"#\\" class=\\"gl-link commit-sha\\"></a></p>
    <p class=\\"gl-m-0\\"><span><a href=\\"http://localhost/group1/project1/-/issues/3\\" class=\\"gl-link gl-text-black-normal issue-date\\"></a></span></p>
  </div>
</div>"
`;
